### ISSUE TITLE
[BUG] Instability after power failure

### ISSUE THREAD

**Issue by a01fe** [longhorn/longhorn#9978]
## Describe the bug

After recovering from a cluster power failure, longhorn pods are frequently OOM killed, and high load average on nodes causes nodes to be not ready.

## Support bundle for troubleshooting

Generating a support bundle via the UI got stuck at 16%, but was able to copy what looks like a complete zip from the support bundle pod.

## Environment

 - Longhorn version: 1.7.2
 - Impacted volume (PV): I've scaled down some workloads attempting to improve stability. As far as I can tell, volumes still attached seem to be okay. 
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): `kubectl`
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: k3s v1.31.2+k3s1
   - Number of control plane nodes in the cluster: 1
   - Number of worker nodes in the cluster: 7
 - Node config
   - OS type and version: debian 12
   - Kernel version: 6.1.0-28-amd64, 6.1.21-v8+
   - CPU per node: 4 nodes - 4 core amd64, 4 nodes - Raspberry PI 4 arm64
   - Memory per node: amd nodes 16GB, arm64 nodes 8GB
   - Disk type (e.g. SSD/NVMe/HDD): SSD
   - Network bandwidth between the nodes (Gbps): 1Gbps
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): baremetal
 - Number of Longhorn volumes in the cluster: 32

**Comment by derekbit** [comment]
@a01fe Which pods experienced the OOM?

cc @mantissahz (community coordinator)

**Comment by a01fe** [comment]
> Did the Longhorn pods have high memory usage?

Not before power failures last week during a storm. Now there lots of these in each node's logs:

```
Dec 15 21:15:16 k3ss01 kernel: Out of memory: Killed process 1854636 (longhorn) total-vm:1993060kB, anon-rss:19824kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:248kB oom_score_adj:999
Dec 15 21:15:16 k3ss01 kernel: Out of memory: Killed process 1854544 (longhorn) total-vm:1992804kB, anon-rss:17336kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:248kB oom_score_adj:999
Dec 15 21:15:16 k3ss01 kernel: Out of memory: Killed process 1954374 (longhorn) total-vm:2066792kB, anon-rss:18208kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:252kB oom_score_adj:999
Dec 15 21:15:16 k3ss01 kernel: Out of memory: Killed process 2056802 (longhorn) total-vm:2066792kB, anon-rss:19444kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:268kB oom_score_adj:999
Dec 15 21:15:16 k3ss01 kernel: Out of memory: Killed process 11709 (longhorn) total-vm:2279860kB, anon-rss:19488kB, file-rss:4kB, shmem-rss:0kB, UID:0 pgtables:384kB oom_score_adj:999
Dec 15 21:15:16 k3ss01 kernel: Out of memory: Killed process 7410 (longhorn-instan) total-vm:9518844kB, anon-rss:3324324kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:10088kB oom_score_adj:999
```

> Can you provide the pods' CPU and Memory metrics over time?

I had scaled down prometheus along with other non-essential workloads to see it would help with stability. I'll scale it back up and and try to get this.

> Which pods experienced the OOM?

longhorn-manager:

```
$ kubectl get -n longhorn-system pod -o json | jq '.items[] | .status.containerStatuses[] | select(.lastState.terminated.reason == "OOMKilled") | .image'
"docker.io/longhornio/longhorn-manager:v1.7.2"
"docker.io/longhornio/longhorn-manager:v1.7.2"
"docker.io/longhornio/longhorn-manager:v1.7.2"
"docker.io/longhornio/longhorn-manager:v1.7.2"
```
> What are your workloads? Any steps to reproduce the case?

At this point, I haven't found any steps to *not* reproduce the problem 😹 I've scaled back as many non-essential workloads as I can, and I still wind up with nodes with load averages > 150.

I'm not sure if this was the case before the "event", but I've noticed that there are *lots* of "Pending" lhb resources.

```
$ kubectl get lhb -n longhorn-system | grep Pending | wc -l
    3697
```

Is this normal?

**Comment by derekbit** [comment]
longhorn-manager seems impossible to use too much memory. Are you using RWX volumes?

**Comment by james-munson** [comment]
Looking through the volumes.yaml, all 31 volumes are `accessMode: rwo`  No RWX, anyway.

One interesting thing is that there are 4,422 backups in backups.yaml.  Also there is a lot of logging in the longhorn-manager logs about this kind of failure:
```
2024-12-14T12:52:16.635079681-08:00 time="2024-12-14T20:52:16Z" level=warning msg="Failed to get the snapshot daily-ba-98b8345e-33e2-4f4b-ae43-aab13ff8671c in volume pvc-facc4660-55ff-42b6-bc5d-c00ad44bf7e2 when syncing backup status" func="controller.(*BackupController).syncBackupStatusWithSnapshotCreationTimeAndVolumeSize" file="backup_controller.go:901" controller=longhorn-backup error="<nil>" node=k3sa02
2024-12-14T12:52:16.636694412-08:00 time="2024-12-14T20:52:16Z" level=warning msg="Waiting to enable the backup monitor for backup backup-0e76a120bcaa4a0c" func="controller.(*BackupController).reconcile" file="backup_controller.go:414" backup=backup-0e76a120bcaa4a0c controller=longhorn-backup error="failed to get the snapshot daily-ba-98b8345e-33e2-4f4b-ae43-aab13ff8671c before enabling backup monitor: snapshot.longhorn.io \"daily-ba-98b8345e-33e2-4f4b-ae43-aab13ff8671c\" not found" node=k3sa02
```

**Comment by a01fe** [comment]
I'm not using RWX volumes.

Here's memory usage on one amd64 node, other amd64 nodes look similar.

<img width="1385" alt="image" src="https://github.com/user-attachments/assets/2e99dd62-12cb-4d60-8ec5-514dd9f210ea" />

Memory usage on an arm64 node:

<img width="1389" alt="image" src="https://github.com/user-attachments/assets/9f73e1df-7ec6-40d8-a878-dec644bc7e38" />

longhorn-manager memory usage:

<img width="680" alt="image" src="https://github.com/user-attachments/assets/0071d44f-8660-4312-ae28-3af40093aa76" />

instance-manager memory usage:

<img width="629" alt="image" src="https://github.com/user-attachments/assets/3438b5af-549b-465c-9617-33125c7558a8" />

**Comment by james-munson** [comment]
Would you pick up a random longhorn-manager and export its pprof heap result?:
1. Forward the pprof port for the longhorn manager:
```
kubectl -n longhorn-system port-forward <longhorn manager pod name> 6060:6060
```
2. Get the heap profiling file:
```
wget -O heap.out "http://localhost:6060/debug/pprof/heap"
```
3.  Upload the profiling result to this ticket. Of course, you can directly check it as well
```
go tool pprof heap.out
(pprof) web
```

**Comment by a01fe** [comment]
Here's one about an hour after restarting all the nodes.
[heap.out.tar.gz](https://github.com/user-attachments/files/18173637/heap.out.tar.gz)

**Comment by a01fe** [comment]
Here's another from a `longhorn-manager` pod using 7G.

[heap-4v5fd.out.tar.gz](https://github.com/user-attachments/files/18185977/heap-4v5fd.out.tar.gz)


```
NAME                     CPU(cores)   MEMORY(bytes)   
longhorn-manager-4v5fd   355m         7052Mi          
longhorn-manager-77cdj   59m          2997Mi          
longhorn-manager-hgtzw   171m         766Mi           
longhorn-manager-qvbc4   27m          2733Mi          
longhorn-manager-s8lnm   36m          1625Mi          
longhorn-manager-sq9vq   56m          1519Mi          
longhorn-manager-srjlz   49m          1267Mi          
longhorn-manager-zqbtf   53m          1414Mi          
```

**Comment by james-munson** [comment]
Sure looks like a repeated GRPC call that is accumulating buffer space and not completing, likely related to the backup monitor that is unable to start.  In the source for longhorn/longhorn-manager, synced to tags/v1.7.2, 

```
ith longhorn/longhorn-manager synced to tags/v1.7.2, do

/repos/Longhorn/longhorn-manager$ go tool pprof --source_path=. <path-to-9978>/heap.out>
File: longhorn-manager
Type: inuse_space
Time: Dec 17, 2024 at 6:49pm (MST)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 493.20MB, 80.00% of 616.50MB total
Dropped 422 nodes (cum <= 3.08MB)
Showing top 10 nodes out of 113
      flat  flat%   sum%        cum   cum%
  193.92MB 31.46% 31.46%   193.92MB 31.46%  bufio.NewReaderSize (inline)
  185.65MB 30.11% 61.57%   185.65MB 30.11%  google.golang.org/grpc/internal/transport.newBufWriter (inline)
   35.10MB  5.69% 67.26%    35.60MB  5.78%  sigs.k8s.io/json/internal/golang/encoding/json.(*decodeState).literalStore
   26.01MB  4.22% 71.48%    26.01MB  4.22%  runtime.malg
   12.51MB  2.03% 73.51%    12.51MB  2.03%  reflect.mapassign0
    9.50MB  1.54% 75.05%   402.08MB 65.22%  google.golang.org/grpc/internal/transport.newHTTP2Client
    8.51MB  1.38% 76.43%     8.51MB  1.38%  reflect.unsafe_NewArray
    8.50MB  1.38% 77.81%     8.50MB  1.38%  google.golang.org/grpc/internal/buffer.NewUnbounded
       7MB  1.14% 78.95%     9.50MB  1.54%  context.(*cancelCtx).propagateCancel
    6.50MB  1.05% 80.00%    15.50MB  2.51%  context.withCancel (inline)
```
And the top two have only grown in the later pprof, up to a combined 1.5 GB
```
~/repos/Longhorn/longhorn-manager$ go tool pprof --source_path=. <path-to-9978>/heap-4v5fd.out

(pprof) top
Showing nodes accounting for 2440.20MB, 85.21% of 2863.92MB total
Dropped 400 nodes (cum <= 14.32MB)
Showing top 10 nodes out of 102
      flat  flat%   sum%        cum   cum%
  779.84MB 27.23% 27.23%   779.84MB 27.23%  google.golang.org/grpc/internal/transport.newBufWriter
  774.67MB 27.05% 54.28%   774.67MB 27.05%  bufio.NewReaderSize
  331.14MB 11.56% 65.84%   331.14MB 11.56%  runtime.malg
  121.01MB  4.23% 70.07%   121.01MB  4.23%  google.golang.org/grpc/internal/buffer.NewUnbounded (inline)
  109.01MB  3.81% 73.87%   109.01MB  3.81%  google.golang.org/grpc/internal/transport.(*controlBuffer).get
   84.01MB  2.93% 76.81%   205.02MB  7.16%  google.golang.org/grpc/internal/grpcsync.NewCallbackSerializer
   81.51MB  2.85% 79.65%    81.51MB  2.85%  context.(*cancelCtx).Done
   68.51MB  2.39% 82.04%   142.51MB  4.98%  context.withCancel
   52.01MB  1.82% 83.86%    77.01MB  2.69%  context.(*cancelCtx).propagateCancel
   38.51MB  1.34% 85.21%    38.51MB  1.34%  golang.org/x/net/http2/hpack.(*headerFieldTable).addEntry
```

**Comment by james-munson** [comment]
![image](https://github.com/user-attachments/assets/0df0e886-9190-4fe1-bf5d-e9a3c805bfb2)

**Comment by james-munson** [comment]
![profile001](https://github.com/user-attachments/assets/328470f3-3e09-46e3-9787-830aa5a9163a)

**Comment by james-munson** [comment]
So it seems clear that there is a leaked GRPC connection in the backup monitoring code, which needs some more poking around.
However, the reason so many efforts are made is because so many backups are pending.  Out of the 4,422, the breakdown by state looks like:
```
grep state: backups.yaml | sort | uniq -c
    682       state: Completed
     21       state: Error
      1       state: "null"
   3718       state: Pending
```
And all the pending ones have a message like this:
```
      messages:
        info: Failed to get the Snapshot daily-ba-d559efa4-d1d2-41b9-8d71-2292a2957b7c
      newlyUploadDataSize: "null"
      ownerID: k3ss01
      progress: 0
      reUploadedDataSize: "null"
      replicaAddress: "null"
      size: "null"
      snapshotCreatedAt: "null"
      snapshotName: "null"
      state: Pending
      url: "null"
```
Where notes about its absence are the only reference to "daily-ba-d559efa4-d1d2-41b9-8d71-2292a2957b7c".  It is not in snapshots.yaml.  That backup and those in similar conditions might as well be deleted because they will never complete.  That will, in turn, lessen the load on memory and CPU as there are fewer backups to try to ask about.  After the cleanup, it may be necessary to restart the longhorn-manager pods to recover the memory.

What happened to snapshot daily-ba-d559efa4-d1d2-41b9-8d71-2292a2957b7c and the others missing?  I don't know.  For the subset I have checked, there are no logs to explain that - only that they are not found.  So that would be the workaround: to delete the Pending backups.  Hopefully the recurring jobs can resume successfully.

**Comment by a01fe** [comment]
I've had a script running for a few days looping through pending backups and `kubectl delete lhb/xxx` them, but it's deleting 10-ish per day, at this rate it'll take months to finish! Is there a better way to delete the pending backups? Also, I'm wondering if it would help to get rid of recurring jobs until I get things cleaned up. Is there way to disable recurring jobs without deleting them all?

**Comment by derekbit** [comment]
@james-munson Assigned the ticket to you. Thank you.

**Comment by james-munson** [comment]
I'm not sure why the deletion would be so slow.  Two possibilities spring to mind:
- The repeated check for backup monitoring is still using a lot of the CPU.
- There is non-trivial work to delete an incremental backup.  See, for instance, the description in https://github.com/longhorn/longhorn/issues/2339#issuecomment-802364652, which still applies, as far as the work to be done.

It probably would be a good idea to stop the ongoing load of the recurring backup job until it catches up.  Are any daily backups succeeding?  One might also consider creating a full backup as a safeguard for some or all volumes, depending on the volume size.  See https://longhorn.io/docs/1.7.2/snapshots-and-backups/backup-and-restore/create-a-backup/#full-backup

FYI, I will be out over the Holidays, so attention to this issue may be sporadic until the New Year.

**Comment by james-munson** [comment]
What's the status on this?  Are pending backups still deleting slowly (and memory still climbing?)

If so, would it be possible to get another pprof dump of heap, and also this time of profile and goroutine?  The command is similar, just replace "heap" with "profile" and "goroutine" in the http path:
```
wget -O profile.out "http://localhost:6060/debug/pprof/profile"
wget -O goroutine.out "http://localhost:6060/debug/pprof/goroutine"
```

**Comment by a01fe** [comment]
Count of pending lhb's have slowly dropped from ~3800 to ~1528. Memory still creeps up on longhorn-manager pods and they eventually get OOM killed (or I delete them before that), but nowhere near as quickly as a few weeks ago. New scheduled backups are working okay.

[goroutine.out.tar.gz](https://github.com/user-attachments/files/18339097/goroutine.out.tar.gz)
[profile.out.tar.gz](https://github.com/user-attachments/files/18339100/profile.out.tar.gz)

**Comment by james-munson** [comment]
Some more insight to the various issues from the added profiling and some improved use of the pprof data:

1. There are thousands of goroutines, all parked, waiting on a select in grpcsync (*CallbackSerializer).run()
   Most of those appear to be created by BackupController syncBackupStatusWithSnapshotCreationTimeAndVolumeSize().
   I'm not sure why the run() never returns.  Is some higher-level code never cancelling the context or timing it
out?

2. The memory is almost entirely alloceted by repeated calls to DeepCopy by Longhorn's DataStore, for VolumeAttachment and Engine CRs, driven from various paths, but all ultimately from BackupController.reconcile().  It seems like some of these calls could use an RO version that does not copy.  The allocation is exacerbated by being repeated by each apparently stuck GPRC call.

3. There is another current issue about the Backup getting stuck in "Pending" forever if the snapshot cannot be found (and some further explanation of how that might happen as well as how to potentially fix it) in https://github.com/longhorn/longhorn/issues/10090.

I will try to upload the latest pprof results as graph images.

To start with, I'll look at the DeepCopy question.

**Comment by james-munson** [comment]
![goroutine](https://github.com/user-attachments/assets/5a3e9eee-4e5c-4ab2-9acb-e2125d6f0a3b)

**Comment by james-munson** [comment]
![profile](https://github.com/user-attachments/assets/a8dde051-3e8b-4ec9-ac3a-e8ba171e78be)

**Comment by james-munson** [comment]
![heap](https://github.com/user-attachments/assets/afcebc0e-25c9-4d2d-84bf-43febcb7b7b5)

**Comment by james-munson** [comment]
I'm slightly off track in the analysis of memory.  The alloc graph shows a colossal rate of deep-copying VolumeAttachments as part of the repeated calls from BackupController.reconcile() polling.  And there is some needless copying that could be optimized.  But that memory is eventually garbage-collected.  
As the inuse graph in https://github.com/longhorn/longhorn/issues/9978#issuecomment-2552620796 shows, the culprit for the gradual increase of memory leadiing to the OOM-killing is the accumulation of GRPC connections that are not closed.  I still have no explanation for that.

**Comment by a01fe** [comment]
Another issue at the time of the repeated power outages, was a minio server outside the cluster (the longhorn  backup target) developed a disk issue where io operations would complete normally, but would take many times longer than normal. That condition lasted for a few days before the disks were replaced.

**Comment by james-munson** [comment]
I think I have the leak.
BackupController.checkMonitor() makes an EngineClientProxy in getBackupTarget().  Once created, EngineClientProxy has to have Close() called on it, or it will leak.  Normally checkMonitor() puts it in to the BackupMonitor it returns, and disableMontor() will close it.  But in the failure case where we fail to get the snapshot, checkMonitor doesn't return the EngineClientProxy.  In fact, there are several failure-related early returns in checkMonitor() after the EngineClientProxy has been created that neither close it nor return it to the caller to close.

**Comment by PhanLe1010** [comment]
> The memory is almost entirely allocated by repeated calls to DeepCopy by Longhorn's DataStore, for VolumeAttachment and Engine CRs, driven from various paths, but all ultimately from BackupController.reconcile(). It seems like some of these calls could use an RO version that does not copy. The allocation is exacerbated by being repeated by each apparently stuck GPRC call.

Maybe we can:
1. Change this one to getengineRO https://github.com/longhorn/longhorn-manager/blob/63a01a72d849115e88805a99a1d3a47c7b7d452d/controller/backup_controller.go#L1005-L1009
2. Also this one https://github.com/longhorn/longhorn-manager/blob/63a01a72d849115e88805a99a1d3a47c7b7d452d/controller/backup_controller.go#L819-L822

`VolumeAttachment` seems have to be deep copied because we are going to edit it inside the backup controller

**Comment by PhanLe1010** [comment]
> I think I have the leak. BackupController.checkMonitor() makes an EngineClientProxy in getBackupTarget(). Once created, EngineClientProxy has to have Close() called on it, or it will leak. Normally checkMonitor() puts it in to the BackupMonitor it returns, and disableMonitor() will close it. But in the failure case where we fail to get the snapshot, checkMonitor doesn't return the EngineClientProxy, nor does it close it itself. In fact, there are several failure-related early returns in checkMonitor() after the EngineClientProxy has been created that neither close it nor return it to the caller to close.

Great finding!

**Comment by derekbit** [comment]
@james-munson Can you help update https://github.com/longhorn/longhorn/issues/9978#issuecomment-2585103222? Thank you.

**Comment by james-munson** [comment]
Hi, @a01fe,
This fix has been committed to Longhorn v1.8.0 and it will be backported to 1.7.3 when it is released.

However, if you are inclined to try it, I have made a build of the longhorn-manager image as of v1.7.2 with this one change added to it.  You can point to `jmunson4suse/longhorn-manager:9978-v1.7.2-test-build` in the longhorn-manager daemonset to run it.
The hope would be that by not piling up all the leaks, it would run faster and not be OOM-killed.

**Comment by PatrickHuetter** [comment]
@james-munson i would also test it since i have 

> jmunson4suse/longhorn-manager:9978-v1.7.2-test-build

I would also like to test the image, as I'm currently experiencing strange behavior in my k3s cluster. Despite having only two small workloads, the RAM usage on the node is extremely high, and the CPU load is as well. This seems to be caused by the k3s server (1.32.0) and partly by Longhorn 1.7.2. I haven’t been able to pinpoint the exact issue yet, but I noticed that Longhorn had problems with S3 and various snapshots. I likely experienced a power outage, and the S3 server might have also been temporarily unavailable. Since I’m already using Longhorn 1.7.2, it shouldn’t be an issue to briefly switch to your image jmunson4suse/longhorn-manager:9978-v1.7.2-test-build, should it?

**Comment by a01fe** [comment]
@james-munson I patched in your test-build but switched back because half my nodes are `arm64` 🤷‍♀️😹

My cluster is almost caught up, it's down to 438 pending `lhb`s!

**Comment by PatrickHuetter** [comment]
@james-munson I'm already running your patched version but in my case (regarding my problem https://github.com/longhorn/longhorn/issues/10189 ) nothing has changed. Also a full node restart afterwards didn't change anything.

**Comment by derekbit** [comment]
@yangchiu Is the ticket validated?

**Comment by yangchiu** [comment]
Verification passed on v1.8.x-head (longhorn-manager baec055) by running the following test case repeatedly:

```
Restart Cluster While Heavy Writing And Recurring Jobs Exist
    [Tags]    recurring_job    cluster
    Given Create volume 0 with    size=2Gi    numberOfReplicas=1    dataEngine=v1
    And Create volume 1 with    size=2Gi    numberOfReplicas=3    dataEngine=v1
    And Create volume 2 with    size=2Gi    numberOfReplicas=3    accessMode=RWX    dataEngine=v1
    And Attach volume 0
    And Attach volume 1
    And Attach volume 2
    And Keep writing data to volume 0
    And Keep writing data to volume 1
    And Keep writing data to volume 2
    And Create snapshot and backup recurringjob for volume 0
    And Create snapshot and backup recurringjob for volume 1
    And Create snapshot and backup recurringjob for volume 2

    FOR    ${i}    IN RANGE    ${LOOP_COUNT}
        When Restart cluster
        And Wait for longhorn ready
        And Wait for volume 0 healthy
        And Wait for volume 1 healthy
        And Wait for volume 2 healthy
        And Check volume 0 works
        And Check volume 1 works
        And Check volume 2 works
    END
```

After over 24 hours of testing (with the cluster being restarted more than 120 times), no significant increase in memory usage was observed, and no longhorn-manager pod was killed due to OOM. 

At the start of the test:
```
# kubectl top nodes
NAME            CPU(cores)   CPU%   MEMORY(bytes)   MEMORY%   
ip-10-0-1-186   155m         3%     971Mi           6%        
ip-10-0-2-207   1144m        28%    974Mi           6%        
ip-10-0-2-227   1502m        37%    798Mi           5%        
ip-10-0-2-42    1447m        36%    870Mi           6%       
 
# kubectl top pods -n longhorn-system
NAME                                                CPU(cores)   MEMORY(bytes)   
csi-attacher-79866cdcf8-7c965                       1m           10Mi            
csi-attacher-79866cdcf8-g4l69                       1m           8Mi             
csi-attacher-79866cdcf8-x9dns                       1m           8Mi             
csi-provisioner-664cb5bdd5-2rv2m                    1m           13Mi            
csi-provisioner-664cb5bdd5-6q88v                    1m           10Mi            
csi-provisioner-664cb5bdd5-lkccj                    1m           10Mi            
csi-resizer-64f6fb4459-9bb5r                        1m           8Mi             
csi-resizer-64f6fb4459-chnvd                        1m           8Mi             
csi-resizer-64f6fb4459-r2s24                        2m           10Mi            
csi-snapshotter-7b7db78f9-4jv7j                     1m           8Mi             
csi-snapshotter-7b7db78f9-dqf2x                     1m           11Mi            
csi-snapshotter-7b7db78f9-phnwb                     1m           10Mi            
engine-image-ei-0b2ebe16-9rnws                      8m           22Mi            
engine-image-ei-0b2ebe16-lchr6                      8m           25Mi            
engine-image-ei-0b2ebe16-xzqp8                      7m           20Mi            
instance-manager-305c0cf808dc82c8c81c9ba92d8f30b6   352m         90Mi            
instance-manager-3cd95a26eace51269266ec9154e03e3a   26m          225Mi           
instance-manager-44b8c593c2c32096da6e91e6c130d991   1010m        40Mi            
instance-manager-606613143728f6bd090ee968727b5684   1008m        38Mi            
instance-manager-94ca66b5b99a45d52e24e0e10908a750   1008m        40Mi            
instance-manager-c99127b5041f570dc2649bd9e6579be7   188m         125Mi           
longhorn-csi-plugin-n7ggf                           1m           30Mi            
longhorn-csi-plugin-nchsq                           1m           30Mi            
longhorn-csi-plugin-p9m9q                           1m           31Mi            
longhorn-driver-deployer-7fc8d96d9f-c9t76           1m           15Mi            
longhorn-manager-lgqlv                              15m          76Mi            
longhorn-manager-nf57z                              61m          99Mi            
longhorn-manager-xn7jp                              8m           101Mi           
longhorn-ui-5698dc78d9-cjdnk                        0m           8Mi             
longhorn-ui-5698dc78d9-rzjdj                        0m           4Mi  
```

After 24 hours of testing and more than 120 cluster restarts:
```
# kubectl top nodes
NAME            CPU(cores)   CPU%   MEMORY(bytes)   MEMORY%   
ip-10-0-1-186   185m         4%     1264Mi          7%        
ip-10-0-2-207   1114m        27%    939Mi           6%        
ip-10-0-2-227   1095m        27%    865Mi           6%        
ip-10-0-2-42    1100m        27%    869Mi           6%        

# kubectl top pods -n longhorn-system
NAME                                                CPU(cores)   MEMORY(bytes)   
csi-attacher-79866cdcf8-7c965                       1m           12Mi            
csi-attacher-79866cdcf8-g4l69                       1m           8Mi             
csi-attacher-79866cdcf8-x9dns                       1m           12Mi            
csi-provisioner-664cb5bdd5-2rv2m                    1m           12Mi            
csi-provisioner-664cb5bdd5-6q88v                    1m           15Mi            
csi-provisioner-664cb5bdd5-lkccj                    1m           10Mi            
csi-resizer-64f6fb4459-9bb5r                        2m           12Mi            
csi-resizer-64f6fb4459-chnvd                        1m           10Mi            
csi-resizer-64f6fb4459-r2s24                        1m           9Mi             
csi-snapshotter-7b7db78f9-4jv7j                     1m           10Mi            
csi-snapshotter-7b7db78f9-dqf2x                     1m           11Mi            
csi-snapshotter-7b7db78f9-phnwb                     1m           12Mi            
engine-image-ei-0b2ebe16-9rnws                      9m           22Mi            
engine-image-ei-0b2ebe16-lchr6                      7m           25Mi            
engine-image-ei-0b2ebe16-xzqp8                      8m           26Mi            
instance-manager-305c0cf808dc82c8c81c9ba92d8f30b6   9m           112Mi           
instance-manager-3cd95a26eace51269266ec9154e03e3a   19m          163Mi           
instance-manager-44b8c593c2c32096da6e91e6c130d991   1011m        40Mi            
instance-manager-606613143728f6bd090ee968727b5684   1010m        40Mi            
instance-manager-94ca66b5b99a45d52e24e0e10908a750   1011m        42Mi            
instance-manager-c99127b5041f570dc2649bd9e6579be7   7m           73Mi            
longhorn-csi-plugin-n7ggf                           1m           36Mi            
longhorn-csi-plugin-nchsq                           1m           33Mi            
longhorn-csi-plugin-p9m9q                           1m           37Mi            
longhorn-driver-deployer-7fc8d96d9f-c9t76           1m           15Mi            
longhorn-manager-lgqlv                              15m          100Mi           
longhorn-manager-nf57z                              15m          108Mi           
longhorn-manager-xn7jp                              5m           82Mi            
longhorn-ui-5698dc78d9-cjdnk                        0m           4Mi             
longhorn-ui-5698dc78d9-rzjdj                        0m           4Mi   
```

[supportbundle_9569533f-586a-4a75-a0c7-153c7e5ae751_2025-01-17T01-23-29Z.zip](https://github.com/user-attachments/files/18447234/supportbundle_9569533f-586a-4a75-a0c7-153c7e5ae751_2025-01-17T01-23-29Z.zip)
